sw/device/tock/chips/opentitan_common/src/gpio.rs

```diff
@@ -0,0 +1,253 @@
+//! General Purpose Input/Output driver.
+
+use tock::kernel as kernel;
+
+use kernel::common::cells::OptionalCell;
+use kernel::common::registers::{
+    register_bitfields, register_structs, Field, ReadOnly, ReadWrite, WriteOnly,
+};
+use kernel::common::StaticRef;
+use kernel::hil::gpio;
+
+register_structs! {
+    pub GpioRegisters {
+        (0x00 => intr_state: ReadWrite<u32, pins::Register>),
+        (0x04 => intr_enable: ReadWrite<u32, pins::Register>),
+        (0x08 => intr_test: WriteOnly<u32, pins::Register>),
+        (0x0c => data_in: ReadOnly<u32, pins::Register>),
+        (0x10 => direct_out: ReadWrite<u32, pins::Register>),
+        (0x14 => masked_out_lower: ReadWrite<u32, mask_half::Register>),
+        (0x18 => masked_out_upper: ReadWrite<u32, mask_half::Register>),
+        (0x1c => direct_oe: ReadWrite<u32, pins::Register>),
+        (0x20 => masked_oe_lower: ReadWrite<u32, mask_half::Register>),
+        (0x24 => masked_oe_upper: ReadWrite<u32, mask_half::Register>),
+        (0x28 => intr_ctrl_en_rising: ReadWrite<u32, pins::Register>),
+        (0x2c => intr_ctrl_en_falling: ReadWrite<u32, pins::Register>),
+        (0x30 => intr_ctrl_en_lvlhigh: ReadWrite<u32, pins::Register>),
+        (0x34 => intr_ctrl_en_lvllow: ReadWrite<u32, pins::Register>),
+        (0x38 => ctrl_en_input_filter: ReadWrite<u32, pins::Register>),
+        (0x3c => @END),
+    }
+}
+
+register_bitfields![u32,
+    pub pins [
+        pin0 0,
+        pin1 1,
+        pin2 2,
+        pin3 3,
+        pin4 4,
+        pin5 5,
+        pin6 6,
+        pin7 7,
+        pin8 8,
+        pin9 9,
+        pin10 10,
+        pin11 11,
+        pin12 12,
+        pin13 13,
+        pin14 14,
+        pin15 15,
+        pin16 16,
+        pin17 17,
+        pin18 18,
+        pin19 19,
+        pin20 20,
+        pin21 21,
+        pin22 22,
+        pin23 23,
+        pin24 24,
+        pin25 25,
+        pin26 26,
+        pin27 27,
+        pin28 28,
+        pin29 29,
+        pin30 30,
+        pin31 31
+    ],
+    mask_half [
+        data OFFSET(0) NUMBITS(16) [],
+        mask OFFSET(16) NUMBITS(16) []
+    ]
+];
+
+pub struct GpioPin {
+    registers: StaticRef<GpioRegisters>,
+    pin: Field<u32, pins::Register>,
+    client: OptionalCell<&'static dyn gpio::Client>,
+}
+
+impl GpioPin {
+    pub const fn new(base: StaticRef<GpioRegisters>, pin: Field<u32, pins::Register>) -> GpioPin {
+        GpioPin {
+            registers: base,
+            pin: pin,
+            client: OptionalCell::empty(),
+        }
+    }
+
+    #[inline(always)]
+    fn half_set(
+        val: bool,
+        field: Field<u32, pins::Register>,
+        lower: &ReadWrite<u32, mask_half::Register>,
+        upper: &ReadWrite<u32, mask_half::Register>,
+    ) {
+        let shift = field.shift;
+        let bit = if val { 1u32 } else { 0u32 };
```

Comment:
Unlike in `C`, you don't need to put `u32` everywhere to make numbers unsigned. The language will pick the appropriate type based on usage, which works out for numbers.